File: ruby/6k/ball-upwards/ball.rb
#!/usr/bin/env ruby

def max_ball(v)
  time = 0.0
  last_height = 0
  increment = 0.1
  while true do
    time += increment
    this_height = height(v, time)
    if this_height < last_height
      return ((time - increment) * 10).round
    end
    last_height = this_height
  end
  
end

def height(v_km_h, t)
  v_km_s = v_km_h.to_f / (60 * 60).to_f
  v_m_s = v_km_s.to_f * (1000.0)
  g = 9.81
  h = v_m_s*t - (0.5*g*t*t)
  return h
end

puts max_ball(15)
puts max_ball(25)
puts max_ball(37)


#h = v*t - 0.5*g*t*t
#    (m/s) * s      m/s^2 * s * s
#        m                  m

File: elixir/6k/ball-upwards/ball.exs
#! /usr/bin/env elixir

defmodule Maxball do
  # h = v*t - (0.5*g*t*t)
  # h' = v - g*t
  # h' = 0 = v - g*t
  # g*t = v
  # t = v / g
  @g 9.81
  def max_ball(v_km_h) do
    (kmh_to_ms(v_km_h) / @g * 10)
    |> Float.round()
  end

  def kmh_to_ms(x), do: x / (60.0 * 60.0) * 1000.0
end

[15, 25, 37]
|> Enum.map(fn x -> Maxball.max_ball(x) end)
|> IO.inspect()
